content/en/blog/2023/php-auto-instrumentation/index.md

```diff
@@ -0,0 +1,169 @@
+
+## Background
+
+[Observability API](https://www.datadoghq.com/blog/engineering/php-8-observability-baked-right-in/)
+allows register and execute additional code (function) before and after original
+one without introducing additional performance penalties in other areas, so in
+other words, we pay only for what we use and only for altered function. Before
+PHP 8, the most common technique for adding tracing capabilities was altering
+`zend_execute_ex` function (a monkey patching kind technique), however this
+could lead to stack overflow and performance problems as whole application paid
+for that. There were also considered other approaches like compile time AST
+modifications which seems feasible, howerer there are no known production ready
+tracers that works in this way up to date.
+
+## Observability API from auto-instrumentation perspective
+
+At the moment of this writing,
+[observability API](https://github.com/php/php-src/blob/PHP-8.0/Zend/zend_observer.h)
+is used by
+[c extension](https://github.com/open-telemetry/opentelemetry-php-instrumentation)
+and exposes one function with following interface:
+
+```php
+function hook(
+    ?string $class,
+    string $function,
+    ?\Closure $pre = null,
+    ?\Closure $post = null,
+): bool {}
+```
+
+This function can be used from user application in order to add additional
+functionality executed before and after. Below code snippet shows how to
+instrument dummy `helloworld` function:
+
+```php
+
+function helloWorld() {
+  echo 'helloWorld';
+}
+
+\OpenTelemetry\Instrumentation\hook(null, 'helloWorld',
+    static function (?string $class, array $params, ?string $classname, string $functionname, ?string $filename, ?int $lineno)
+    {
+      echo 'before';
+    },
+    static function (mixed $object, array $params, mixed $return, ?Throwable $exception)
+    {
+      echo 'after';
+    }
+);
+```
+
+In the same way, we implemented tracing support for the most important
+`interfaces/libraries/frameworks` that are parts of
+[Contrib](https://github.com/open-telemetry/opentelemetry-php-contrib/tree/main/src/Instrumentation)
+repo. Each `auto-instrumentation` package uses above `hook` function in order to
+register and provide tracing functionality. One missing thing, not mentioned yet
+is an `API and SDK` used to create traces and other necessary components. This
+is content and responisibility of
+[main](https://github.com/open-telemetry/opentelemetry-php) repo which is
+foundation for everything.
+
+![php-rel](php-rel.png)
+
+## How to use it
+
+All components necessary for auto-instrumentation can be installed manually,
+however we invested time to lower the barier and created an installer that can
+do that for you. This section will show how auto-instrument simple php `laravel`
+application created from scratch.
+
+First step is to create a demo application. Here we use the popular
+[laravel](https://laravel.com/docs/10.x/installation). framework:
+
+```
```

Comment:
Add the `sh` language to all command-line code blocks in this page, like this:

```suggestion
```sh
```